src/System.Security.Permissions/src/System/Security/SecurityElement.cs

```diff
@@ -2,30 +2,827 @@
+                {
+                    string strAttrValue = (string)m_lAttributes[i + 1];
+
+                    return Unescape(strAttrValue);
+                }
+            }
+
+            // In the case where we didn't find it, we are expected to
+            // return null
+            return null;
+        }
+
+        public SecurityElement SearchForChildByTag(string tag)
+        {
+            // Go through all the children and see if we can
+            // find the one are are asked for (matching tags)
+            if (tag == null)
+                throw new ArgumentNullException(nameof(tag));
+
+            // Note: we don't check for a valid tag here because
+            // an invalid tag simply won't be found.    
+            if (m_lChildren == null)
+                return null;
+
+            IEnumerator enumerator = m_lChildren.GetEnumerator();
+
+            while (enumerator.MoveNext())
+            {
+                SecurityElement current = (SecurityElement)enumerator.Current;
+
+                if (current != null && string.Equals(current.Tag, tag))
+                    return current;
+            }
+            return null;
+        }
+
+        internal string SearchForTextOfLocalName(string strLocalName)
+        {
+            // Search on each child in order and each
+            // child's child, depth-first
+            if (strLocalName == null)
+                throw new ArgumentNullException(nameof(strLocalName));
+
+            // Note: we don't check for a valid tag here because
+            // an invalid tag simply won't be found.    
+            if (m_strTag == null)
+                return null;
+            if (m_strTag.Equals(strLocalName) || m_strTag.EndsWith(":" + strLocalName, StringComparison.Ordinal))
+                return Unescape(m_strText);
+            if (m_lChildren == null)
+                return null;
+
+            IEnumerator enumerator = m_lChildren.GetEnumerator();
+
+            while (enumerator.MoveNext())
+            {
+                string current = ((SecurityElement)enumerator.Current).SearchForTextOfLocalName(strLocalName);
+
+                if (current != null)
+                    return current;
+            }
+            return null;
+        }
+
+        public string SearchForTextOfTag(string tag)
+        {
+            // Search on each child in order and each
+            // child's child, depth-first
+            if (tag == null)
+                throw new ArgumentNullException(nameof(tag));
+
+            // Note: we don't check for a valid tag here because
+            // an invalid tag simply won't be found.    
+            if (string.Equals(m_strTag, tag))
+                return Unescape(m_strText);
+            if (m_lChildren == null)
+                return null;
+
+            IEnumerator enumerator = m_lChildren.GetEnumerator();
+
+            ConvertSecurityElementFactories();
+
+            while (enumerator.MoveNext())
```

Comment:
Nit: `foreach`... but, I see the enumerator is retrieved then `ConvertSecurityElementFactories()` is called... If `ConvertSecurityElementFactories()` changes the ArrayList, isn't the enumerator going to throw because of a version mismatch? Maybe that behavior is intended?